@xen-orchestra/proxy/src/app/mixins/reverseProxy.mjs

```diff
@@ -0,0 +1,117 @@
+  const res = new URL(base)
+  const relativeUrl = new URL(relative, base)
+  res.pathname = relativeUrl.pathname
+  relativeUrl.searchParams.forEach((value, name) => {
+    // we do not allow to modify params already specified by config
+    if (!res.searchParams.has(name)) {
+      res.searchParams.append(name, value)
+    }
+  })
+  res.hash = relativeUrl.hash.length > 0 ? relativeUrl.hash : res.hash
+  return res
+}
+export default class ReverseProxy {
+  constructor(app, { httpServer }) {
+    this._app = app
+    httpServer.on('request', this.proxy)
+    httpServer.on('upgrade', this.upgrade)
+  }
+
+  localToBackendUrl(basePath, target, localPath) {
+    let localPathWithoutBase = removeSlash(localPath).substring(basePath.length)
+    localPathWithoutBase = './' + removeSlash(localPathWithoutBase)
+    const url = mergeUrl(localPathWithoutBase, target)
+    return url
+  }
+
+  backendToLocalPath(basePath, target, backendUrl) {
+    // keep redirect url relative to local server
+    const localPath = `${basePath}/${backendUrl.pathname.substring(target.pathname.length)}${backendUrl.search}${
+      backendUrl.hash
+    }`
+    return localPath
+  }
+
+  _getConfigFromRequest(req) {
+    for (const [path, config] of Object.entries(this._app.config.get('reverseProxies'))) {
+      const fullPath = '/proxy/v1/' + removeSlash(path)
+      if (req.url.startsWith(fullPath + '/')) {
+        if(typeof config === 'string'){
+          return {
+            path: fullPath,
+            target : config,
+            options: {}
+          }
+        }
+
+        const { target, ...options } = config
+        return {
+          path: fullPath,
+          target,
+          options,
+        }
+      }
+    }
+  }
+
+  proxy(req, res) {
+    const config = this._getConfigFromRequest(req)
+
+    if (config === undefined) {
+      return
+    }
+
+    const url = new URL(config.target)
+    const targetUrl = this.localToBackendUrl(config.path, url, req.originalUrl || req.url)
+    proxy.web(req, res, {
+      ...urlToHttpOptions(targetUrl),
+      ...config.options,
+      onRes: (req, res, proxyRes) => {
+        // rewrite redirect to pass through this proxy
+        if (proxyRes.statusCode === 301 || proxyRes.statusCode === 302) {
+          // handle relative/ absolute location
+          const redirectTargetLocation = new URL(proxyRes.headers.location, url)
+
+          // this proxy should only allow communication between known hosts. Don't open it too much
+          if (redirectTargetLocation.hostname !== url.hostname || redirectTargetLocation.protocol !== url.protocol) {
+            throw new Error(`Can't redirect from ${url.hostname} to ${redirectTargetLocation.hostname} `)
+          }
+          res.writeHead(proxyRes.statusCode, {
+            ...proxyRes.headers,
+            location: this.backendToLocalPath(config.path, url, redirectTargetLocation),
+          })
+          res.end()
+          return
+        }
```

Comment:
Why did you chose to handle redirections?

Was this supported by `xo-server`'s implementation?

Reply:
IMHO, redirections seems to be a basic needs for a proxy, but if they're not needed it can simplify the code greatly. 

httpProxy does not seems to handle redirect by default, and there is no flag `followRedirects` 

What should I do ?

Reply:
I agree with you, let's keep the path redirection :slightly_smiling_face: